pkg/controller/jwks/jwks.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package jwks handles returning JSON encoded information about the
+// server's encryptionn keys. See https://tools.ietf.org/html/rfc75170
+// for information about the server.
+package jwks
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/google/exposure-notifications-server/pkg/logging"
+	"github.com/google/exposure-notifications-verification-server/pkg/cache"
+	"github.com/google/exposure-notifications-verification-server/pkg/database"
+	"github.com/google/exposure-notifications-verification-server/pkg/keyutils"
+	"github.com/google/exposure-notifications-verification-server/pkg/render"
+	"github.com/gorilla/mux"
+	"github.com/rakutentech/jwk-go/jwk"
+
+	"go.uber.org/zap"
+)
+
+// Error codes
+var errBadRealm = errors.New("bad realm")
+var errEncode = errors.New("error encoding key")
+
+// Controller holds all the pieces necessary to show the jwks encoded keys.
+type Controller struct {
+	h        *render.Renderer
+	db       *database.Database
+	keyCache *keyutils.PublicKeyCache
+	cacher   cache.Cacher
+	logger   *zap.SugaredLogger
+}
+
+// HandleIndex returns an http.Handler that handles jwks GET requests.
+func (c *Controller) HandleIndex() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		var keys []*database.SigningKey
+		if err := c.cacher.Fetch(ctx, r.URL.String(), &keys, 5*time.Minute, func() (interface{}, error) {
+			// Grab the URL path components  we need.
+			realmID := mux.Vars(r)["realm"]
+
+			// Find the realm, and the key abstractions from the DB.
+			realm, err := c.getRealm(realmID)
+			if err != nil {
+				return nil, err
+			}
+			var keys []*database.SigningKey
+			keys, err = realm.ListSigningKeys(c.db)
+			if err != nil {
+				return nil, err
+			}
+			return keys, nil
+		}); err != nil {
+			c.renderError(w, err)
+			return
+		}
+
+		// Get the keys.
+		encoded := make([]*jwk.JWK, len(keys))
```

Comment:
does it make sense for this to be the cached entity?

Reply:
Arbitrary to me. I felt the rest of the code was so minimal, and used its own caches that it wasn't a burden. Updated.